### PR TITLE
Drastically improve Xcode 9 provisioning profile detection

### PR DESCRIPTION
**Pull request by KrauseFx**
This PR adds support for merging of multiple profile mappings. Since Xcode 9, you as a user, have to provide the mapping from app identifier to provisioning profile. As we don't want people to have to manually update the config, we can do that automatically for them :) 

screen shot 2017-08-28 at 2 06 54 pm

### CODE REVIEW

**Comment by b3ll** [comment]
:bell:

**Comment by taquitos** [comment]
so drastic

**Comment by b3ll** [comment]
but will it blend?

File: gym/lib/gym/detect_values.rb [hunk 1]
```diff
@@ -121,7 +117,52 @@ def self.detect_selected_provisioning_profiles
...
+      #   1) A provisioning profile is defined in the Xcode project only: we'll take that
+      #   2) A provisioning profile is defined manually from the user: we'll take that (same as 3)
+      #   3) A provisioning profile is defined after calling match: we'll take that (same as 2)
+      #   4) On a conflict (app identifier assigned both in xcode and match)
+      #     4.1) we'll choose whatever matches the `export_method` defined
+      #     4.2) If both include the export_method, we'll prefer the one from match
+      #     4.3) If none has the right export_method, we'll use whatever is defined in the Xcode project
+
+      # 2) is already covered by assigning `hash_to_use` above
+      # 3) see 2
```
Comment (ohayon) [inline gym/lib/gym/detect_values.rb]: These are really great comments ❤️ . I'm not sure I understand the "see 2" here though. 😕

File: gym/lib/gym/detect_values.rb [hunk 2]
```diff
@@ -121,7 +117,52 @@ def self.detect_selected_provisioning_profiles
...
+      provisioning_profile_mapping.each do |bundle_identifier, provisioning_profile|
+        if hash_to_use[bundle_identifier].nil?
+          # 1)
+          hash_to_use[bundle_identifier] = provisioning_profile
+        else
+          # 4) Conflict, we now have to choose if we prefer whatever match provides
+          #    or what's defined in the project
+          if self.app_identifier_contains?(hash_to_use[bundle_identifier], Gym.config[:export_method])
+            # 4.1
+            # 4.2 is automatically covered by this, as we're checking for the one coming from the export plist first
```
Comment (ohayon) [inline gym/lib/gym/detect_values.rb]: Is nothing supposed to happen in this case? Maybe I'm not understanding all of this correctly, I think these comments could be a bit more clear ✨

File: gym/lib/gym/detect_values.rb [hunk 3]
```diff
@@ -121,7 +117,52 @@ def self.detect_selected_provisioning_profiles
...
+          elsif self.app_identifier_contains?(provisioning_profile, Gym.config[:export_method])
+            # 4.1
+            hash_to_use[bundle_identifier] = provisioning_profile
```
Comment (ohayon) [inline gym/lib/gym/detect_values.rb]: `4.1` is mentioned here and in the case above?

File: gym/lib/gym/detect_values.rb (hunk 2 above)
Comment (ohayon) [inline gym/lib/gym/detect_values.rb]: It seems like we only need to check the case on line 149 to do anything right?

File: gym/lib/gym/detect_values.rb [hunk 4]
```diff
@@ -135,6 +176,14 @@ def self.detect_selected_provisioning_profiles
       end
     end
 
+    # Helper method to remove "-" and " " and downcase app identifier
+    # and compare if an app identifier includes a certain string
+    # We do some `gsub`bing, because we can't really know the profile type, so we'll just look at the name and see if it includes
+    # the export method (which it usually does, but with different notations)
+    def self.app_identifier_contains?(str, contains)
+      return str.to_s.gsub("-", "").gsub(" ", "").downcase
```
Comment (ohayon) [inline gym/lib/gym/detect_values.rb]: Should this return a boolean as it ends with a `?`?

File: gym/lib/gym/error_handler.rb [hunk 5]
```diff
@@ -250,7 +250,7 @@ def print_environment_information
 
               next if matching_type.to_s == selected_export_method
               UI.message("")
-              UI.error("There is a mismatch between your provided `export_method` in gym")
+              UI.error("There seems to be a mismatch between your provided `export_method` in gym")
```
Comment (ohayon) [inline gym/lib/gym/error_handler.rb]: ❤️

**Review by ohayon — COMMENTED** [review]
This looks like a great change @KrauseFx. Some of the comments and conditional logic are a bit confusing, but maybe I just don't fully understand the use case for all of this.

One other comment is, it looks like this method is now over 100 lines long (including comments obviously), is there anything we can do to extract some things out into more digestible pieces? 🚀

File: gym/lib/gym/detect_values.rb [hunk 6]
```diff
@@ -121,7 +117,52 @@ def self.detect_selected_provisioning_profiles
       return if provisioning_profile_mapping.count == 0
 
       Gym.config[:export_options] ||= {}
-      Gym.config[:export_options][:provisioningProfiles] = provisioning_profile_mapping
+      hash_to_use = Gym.config[:export_options][:provisioningProfiles].dup || {} # dup so we can show the original values in `verbose` mode
```
Comment (thasegaw) [inline gym/lib/gym/detect_values.rb]: If the 'provisioningProfiles' key does not exist, an error will occur.

File: gym/lib/gym/detect_values.rb [hunk 7]
```diff
@@ -121,7 +117,52 @@ def self.detect_selected_provisioning_profiles
...
+
+      # Now it's time to merge the (potentially) existing `provisioningProfiles` of the `export_options`
+      # with the hash we just created. Both might include information about what profile to use
+      # This is important, as :bell: (some users) didn't know to call match for each app target he wants profiles for
+      # before adding this code, we'd only either use whatever we get from match, or what's defined in the Xcode project
+      # With the code below, we'll make sure to take the best of it:
+      #
+      #   1) A provisioning profile is defined in the Xcode project only: we'll take that
+      #   2) A provisioning profile is defined manually from the user: we'll take that (same as 3)
+      #   3) A provisioning profile is defined after calling match: we'll take that (same as 2)
```
Comment (nafu) [inline gym/lib/gym/detect_values.rb]: If I understand correctly, it looks like

```
1) A provisioning profile is defined (we'll take that)
  1.1) In the Xcode project only
  1.2) Manually from the user
  1.3) After calling match
2) On a conflict (app identifier assigned both in xcode and match)
  2.1) we'll choose whatever matches the `export_method` defined
  2.2) If both include the export_method, we'll prefer the one from match
  2.3) If none has the right export_method, we'll use whatever is defined in the Xcode project
```

File: gym/lib/gym/detect_values.rb [hunk 8]
```diff
@@ -121,7 +117,52 @@ def self.detect_selected_provisioning_profiles
       return if provisioning_profile_mapping.count == 0
 
       Gym.config[:export_options] ||= {}
-      Gym.config[:export_options][:provisioningProfiles] = provisioning_profile_mapping
+      hash_to_use = Gym.config[:export_options][:provisioningProfiles].dup || {} # dup so we can show the original values in `verbose` mode
+
+      # Now it's time to merge the (potentially) existing `provisioningProfiles` of the `export_options`
+      # with the hash we just created. Both might include information about what profile to use
+      # This is important, as :bell: (some users) didn't know to call match for each app target he wants profiles for
```
Comment (nafu) [inline gym/lib/gym/detect_values.rb]: Hi @b3ll 👋

File: gym/lib/gym/detect_values.rb (hunk 4 above)
Comment (nafu) [inline gym/lib/gym/detect_values.rb]: `include?(contains)` missing?

File: gym/lib/gym/error_handler.rb (hunk 5 above)
Comment (nafu) [inline gym/lib/gym/error_handler.rb]: 💚

**Review by nafu — COMMENTED** [review]
This looks great @KrauseFx 👍

File: gym/lib/gym/detect_values.rb (hunk 1 above)
Comment (KrauseFx) [inline gym/lib/gym/detect_values.rb]: It's a reference to the fact that 2 and 3 are the same in this case, I'll update it to make it more clear

File: gym/lib/gym/detect_values.rb (hunk 4 above)
Comment (KrauseFx) [inline gym/lib/gym/detect_values.rb]: oh yessss, whops, what was I thinking...

File: gym/lib/gym/error_handler.rb (hunk 5 above)
Comment (KrauseFx) [inline gym/lib/gym/error_handler.rb]: 💙

**Comment by KrauseFx** [comment]
I'll address the feedback later today / tomorrow morning

File: gym/lib/gym/error_handler.rb (hunk 5 above)
Comment (hjanuschka) [inline gym/lib/gym/error_handler.rb]: 💜

**Comment by fastlane-bot-helper** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by KrauseFx** [comment]
Alright, finally finished this PR up, it's its own class now, lots of tests and more comments, please let me know what you all think 👍

File: gym/lib/gym/code_signing_mapping.rb [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                provisioning_profile_mapping[bundle_identifier] = provisioning_profile_specifier
+              elsif provisioning_profile_uuid.to_s.length > 0
+                provisioning_profile_mapping[bundle_identifier] = provisioning_profile_uuid
+              end
+            end
+          end
+        rescue => ex
+          # We catch errors here, as we might run into an exception on one included project
+          # But maybe the next project actually contains the information we need
+          if Helper.xcode_at_least?("9.0")
```
Comment (ohayon) [inline gym/lib/gym/code_signing_mapping.rb]: If Xcode is < 9, should we re-raise the error here?

**Review by ohayon — APPROVED** [review]
This all looks 💯  and the code is great. I just have a few questions in there about the comments. 🚀

File: gym/lib/gym/code_signing_mapping.rb [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+
+      # Now it's time to merge the (potentially) existing `provisioningProfiles` of the `export_options`
+      # with the hash we just created. Both might include information about what profile to use
+      # This is important, as :bell: (some users) didn't know to call match for each app target he wants profiles for
+      # before adding this code, we'd only either use whatever we get from match, or what's defined in the Xcode project
+      # With the code below, we'll make sure to take the best of it:
+      #
+      #   1) A provisioning profile is defined in the Xcode project only: we'll take that
+      #   2) A provisioning profile is defined manually from the user: we'll take that (same as 3)
+      #   3) A provisioning profile is defined after calling match: we'll take that (same as 2)
```
Comment (KrauseFx) [inline gym/lib/gym/code_signing_mapping.rb]: Basically inside https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/gym.rb we set it, therefore it's the same source as if the user would set it manually. But while linking this I noticed that [this line](https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/gym.rb#L25) would overwrite the user's mapping if the user provided a manual mapping and uses _match_. I'll update the PR 👍

File: gym/lib/gym/code_signing_mapping.rb [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      #   4) On a conflict (app identifier assigned both in xcode and match)
+      #     4.1) we'll choose whatever matches what's defined in the `export_method`
+      #     4.2) If both include the export_method, we'll prefer the one from match
+      #     4.3) If none has the right export_method, we'll use whatever is defined in the Xcode project
+
+      # 2) is already covered by assigning `final_mapping` above
+      # 3) see 2
```
Comment (KrauseFx) [inline gym/lib/gym/code_signing_mapping.rb]: 2 and 3 are the same thing, as it's treated the same way, see comment https://github.com/fastlane/fastlane/pull/9952#discussion_r132236924. I'll update the comments to make it more clear and explain more

File: gym/lib/gym/code_signing_mapping.rb (hunk 9 above)
Comment (KrauseFx) [inline gym/lib/gym/code_signing_mapping.rb]: No, in general, everything below Xcode 9 we don't want to show / raise anything, as no provisioning profile mapping is required to build, so we'd then potentially fail the build for existing setups

**Review by ohayon — APPROVED** [review]

**Comment by taquitos** [comment]
[Drastically merged]

**Pull request merged**